McGonagall Docs Re-Org / MVP Generator (#971)
UnitedIncome/components: components/templates/McGonagall/McGonagall.mdx

# McGonagall

> McGonagall is a question flow framework which uses [xState](https://xstate.js.org/docs/). It's used to generate a large set of data in the form of a questionaire.

Reviewer: ```suggestion
McGonagall is a question flow framework that uses [xState](https://xstate.js.org/docs/). It's used to generate a large set of data in the form of a questionnaire.
```

## Example 🚀

> You can generate a new McGonagall flow by running the following command within this library and answering a few basic questions. Although not neccersary for all implementations, the generateor will assumes you're using Redux in your project.

Reviewer 2: s/neccersary/necessary/

> ```bash

Reviewer 2: Nit: I'd stick with `sh` over `bash` unless you're using `bash`-specific features.

yarn mcgonagall
```

Once the process is complete you'll be presented with a flow that you can start to use as soon as you fix all of the relative imports.

```markdown
├── Flow
│ ├── Flow.js - A parent component that renders the McGonagall component and passes in the necessary information.
│ ├── Flow.cards.js - Includes the card render functions and the steps object.
│ ├── Flow.xstate.js - Includes the stateConfig and the stateOptions.
│ └── Flow.constants.js - Includes an outputs constant (to help keep output names consistent across the files) and other constants needed for the flow.
```

The following sections will go over the `stateConfig`, `stateOptions`, `stateContext` and `steps` props which are used to navigate users through the question flow.

```javascript
<McGonagall
  name="Question Flow"
  stateConfig={stateConfig}
  stateContext={stateContext}
  stateOptions={stateOptions}
  steps={steps}
/>
```

### stateContext 💡

> The `stateContext` prop is used to setup the initial McGonagall state. You can use this prop to pass in methods or pieces of data that are sourced from outside of the question flow.

Reviewer: ```suggestion
The `stateContext` prop is used to setup the initial McGonagall state. You can use this prop to pass in functions or pieces of data that are sourced from outside of the question flow.
```

> In the following example `futureOrCurrent` is not asked in the flow, but the information is used to influence questions within the flow. With this setup anytime you have access to McGonagall's state you'll be able to reference `state.futureOrCurrent`. This can be useful if you need to pass in personalized information to reference within a question, for example a users name or something similar.

Reviewer: ```suggestion
In the following example `futureOrCurrent` is not asked in the flow, but the information is used to influence questions within the flow. With this setup anytime you have access to McGonagall's state you'll be able to reference `state.futureOrCurrent`. This can be useful if you need to pass in personalized information to reference within a question, for example a user's name or something similar.
```

```javascript
/** PropertyEntry Component */
class Flow extends PureComponent {
  /** @inheritdoc */
  render() {
    const stateContext = {
      [outputs.futureOrCurrent]: this.props.match.params.screen,
    };

    return (
      <McGonagall
        name="Add a Property"
        browserHistory={browserHistory}
        exitLocation={FPE_PROPERTIES}
        location={this.props.location}
        stateConfig={stateConfig}
        stateContext={stateContext}
        steps={steps}
        isFetching={this.props.isFetching}
      />
    );
  }
}
```

### stateConfig 💡

> The `stateConfig` prop is where you'll configure the transitions between each step in the question flow, this should be imported from the `Flow.xstate.js` file. Here you can add logic to determine which question is presented after the current one, and logic that

Reviewer: ```suggestion
The `stateConfig` prop is where you'll configure the transitions between each step in the question flow, this should be imported from the `Flow.xstate.js` file. Here you can add logic to determine which question is presented after the current one and logic that
```

> ensures that certain information is present before the user will be allowed to advance.

Reviewer: weird line break

};
```

> In an event for a state node, you can pass in an array of strings that correspond to the name of the action to excute. In the below example, there are 2 potential state nodes a user can go from the `name` state node. The state machine will first check if it passes the condition for the first object in the array so if the `hasNameAndIsMarried` guard returns true, the flow will transition to the owner step and run the 5 actions that are listed. If `hasNameAndIsMarried` returns false, it will then check the next condition. A state node should be able to pass at least one of the conditions.

Reviewer: ```suggestion
In an event for a state node, you can pass in an array of strings that correspond to the name of the action to execute. In the below example, there are 2 potential state nodes a user can go to from the `name` state node. The state machine will first check if it passes the condition for the first object in the array; so if the `hasNameAndIsMarried` guard returns true, the flow will transition to the owner step and run the 5 actions that are listed. If `hasNameAndIsMarried` returns false, it will then check the next condition. A state node should be able to pass at least one of the conditions.
```

};
```

It's best to use descriptive names for your guards and create seperate guards for functionally different conditions so that if you're viewing it in the [xState visualizer](https://xstate.js.org/viz/), you can easily understand it.

> When the McGonagall component first mounts, the framework uses the guards to rehydrate the flow. It tries transitioning at each step until it hits a step that it can't navigate away from because it doesn't satisfy any condition. It's important to add guards to each State Node possible transition so that it the flow starts back up at the appropriate place.

Reviewer: ```suggestion
When the McGonagall component first mounts, the framework uses the guards to rehydrate the flow. It tries transitioning at each step until it hits a step that it can't navigate away from because it doesn't satisfy any condition. It's important to add guards to each State Node possible transition so that the flow starts back up at the appropriate place.
```

      'Explain why we need this. Giving this property a name helps you identify this asset when viewing it in your financial plan.  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
    shortTitle: 'Name',
    outputs: [outputs.nickname],
    card: renderName,
    clearFuture: false, // Boolean to indicate whether all future cards needs to be cleared when editing thihs card
  },
];
```

#### steps.card

The function in the card property will get passed 2 arguments, props and state.

> The props argument will include information from the Mcgonagall framework and was designed so you can pass it straight through to the cards:

Reviewer 2: s/Mcgonagall/McGonagall/

];
```

## Visualizing the Flow 💡

> If you'd like to visualize your McGonnagal flow you can copy and paste the `stateConfig` and `stateOptions` objects into the [xState visualizer tool](https://statecharts.github.io/xstate-viz/) and pass both of the variables into `Machine` like so.

Reviewer 2: s/McGonnagal/McGonagall/

    },
  },
  guards: {
    hasNameAndIsMarried: (ctx, eve) => true,
    hasNameAndIsNotMarried: (ctx, eve) => true,
  },
};

const fetchMachine = Machine(stateConfig, stateOptions);
```

> This will produce the following, allowing you to see how each question transitions and affects eachother.

Reviewer 2: s/eachother/each other/

import {saveToStateContext} from '@unitedincome/components/dist/mcgonagall';
```

### Typography Styles

> There are a number of typograhy utility classes that be found [here](/?path=/story/utilities-styles--typography).

Reviewer 2: s/typograhy/typography/
